matter_server/server/storage.py

```diff
@@ -0,0 +1,142 @@
+
+from genericpath import isfile
+
+from ..common.helpers.json import JSON_DECODE_EXCEPTIONS, json_dumps, json_loads
+
+DEFAULT_SAVE_DELAY = 30
+
+if TYPE_CHECKING:
+    from .server import MatterServer
+
+
+class StorageController:
+    """Controller that handles storage of persistent data."""
+
+    def __init__(self, server: "MatterServer") -> None:
+        """Initialize storage controller."""
+        self.server = server
+        self.logger = logging.getLogger(__name__)
+        self._data: Dict[str, Any] = {}
+        self._timer_handle: asyncio.TimerHandle = None
+
+    @property
+    def filename(self) -> str:
+        """Return full path to (fabric-specific) storage file."""
+        return os.path.join(
+            self.server.storage_path,
+            f"{self.server.device_controller.compressed_fabric_id}.json",
+        )
+
+    async def start(self) -> None:
+        """Async initialize of controller."""
+        await self._load()
+        self.logger.debug("Started.")
+
+    async def stop(self) -> None:
+        """ "Handle logic on server stop."""
+        if not self._timer_handle:
+            # no point in forcing a save when there are no changes pending
+            return
+        await self.async_save()
+        self.logger.debug("Stopped.")
+
+    def get(self, key: str, default: Any = None, subkey: str | None = None) -> Any:
+        """Get data from specific (sub)key."""
+        if subkey:
+            # we provide support for (1-level) nested dict
+            return self._data.get(key, {}).get(subkey, default)
+        return self._data.get(key, default)
+
+    def set(
+        self,
+        key: str,
+        value: Any,
+        subkey: str | None = None,
+        force: bool = False,
+    ) -> None:
+        """Set a (sub)value in persistent storage."""
+        if not force and self.get(key, subkey=subkey) == value:
+            # no need to save if value did not change
+            return
+        if subkey:
+            # we provide support for (1-level) nested dict
+            self._data.setdefault(key, {})
+            self._data[key][subkey] = value
+        else:
+            self._data[key] = value
+        self.save(force)
+
+    def __getitem__(self, key: str) -> Any:
+        """Get data from specific key."""
+        return self._data[key]
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        """Set a value in persistent storage."""
+        self.set(key, value)
+
+    async def _load(self) -> None:
+        """Load data from persistent storage."""
+        assert not self._data, "Already loaded"
+
+        def _load() -> dict:
+            for filename in self.filename, f"{self.filename}.backup":
+                try:
+                    _filename = os.path.join(self.server.storage_path, filename)
+                    with open(filename, "r") as _file:
+                        data = json_loads(_file.read())
+                        self.logger.debug(
```

Comment:
Maybe move the log and return to an `else:` block?

Reply:
look like some leftover debug